### DISCUSSION TITLE
[Enhancement] Built-in unified Bluetooth Low Energy (BLE) APIs

### DISCUSSION THREAD

**Discussion by LITTOMA** [dotnet/maui#277] [General]
In Xamarin Forms, there is no built-in unified BLE APIs.
If I want to write an APP uses BLE functions, I have to write platform-specific code (or using a 3rd party plugin).
BLE is a fundamental function, and as a multi-platform framework,  I think MAUI should have unified APIs for it.

**Comment by GeraudFabien** [comment]
Hello,
BLE = Bluetooth Low Energy ?

**Reply by bmacombe** [reply]
If so, this request probably belongs in Essentials

**Reply by LITTOMA** [reply]
Yes, exactly.
Edited the title.

**Reply by sultanAalyami** [reply]
be best

**Comment by sultanAalyami** [comment]
![image](https://user-images.githubusercontent.com/12185541/100543249-0eb62780-3260-11eb-9636-11e2eeeaf53a.png)

**Comment by davidortinau** [comment]
BLE is not on our roadmap for .NET 6. We are discussing bringing Essentials together with Forms in dotnet/maui as there's a growing overlap and interdependence.

Still, BLE is such a complex topic I'm not sure it'll ever be something we bake into this product. [Shiny](https://github.com/shinyorg/shiny) seems to be filling this need presently.

**Reply by nitinjs** [reply]
@davidortinau q1. shiny is compatible with .net maui? will it work seamlessly across windows, android, mac and ios?

q2. All xamarin forms libraries are compatible with .net maui??

**Reply by nitinjs** [reply]
@davidortinau BLE is quite common thing across all the platforms, I guess it should be included in .net roadmap.

**Reply by yshteinm** [reply]
It would be nice to have BLE included somewhere for all platforms.

**Reply by Rowingwolf** [reply]
I've tried to use Shiny for Maui and haven't figured it out.  I will say the Shiny documentation is a bit hard to follow.  Maybe if I knew what I was doing it would be pretty clear....but then I probably wouldn't need the documentation!

**Reply by SaddamBInSyed** [reply]
@Rowingwolf 
Did you integrate the Bluetooth part in your, NET MAUI Android platform?

@davidortinau 

Can you please share the latest Document about the Bluetooth integration on MAUI android platform

Thanks

**Reply by gfmoore** [reply]
BLE is an essential requirement. Many devices such as in sports and elsewhere use Bluetooth. It is not a new technology. Yes it's difficult, but for a company like MS... Think too that even AppInventor 2 supports it, Flutter supports it, ReactNative supports it.  I find it astonishing that it's not on "the roadmap"!

**Reply by yshteinm** [reply]
Completely agree

**Reply by LITTOMA** [reply]
MAUI would never be "Multiplatform" until it has a unified BLE API.

**Reply by nexxuno** [reply]
I've spent many many hours on the predecessor of shiny and when you understood how to make it work it was good enough. If this is how MAUI is going to support BT (with a one man project, difficult to use if you are not into Rx programming and super poor documentation) then it's clear MS doesn't want to really invest in the mobile battle. The guy also offers paid consultancy about his library so I guess docs will never be that good... One path would be give that man some good money and a team.

EDIT: I have to say that the docs for Shiny now look good, didn't try this new version though so I can't say if it works better than the previous one and if the docs are aligned with the code

**Reply by LITTOMA** [reply]
> 

The Rx pattern is not a good design for BLE functions. I won't to give it a try in my projects.

**Reply by nexxuno** [reply]
> > 
> 
> The Rx pattern is not a good design for BLE functions. I won't to give it a try in my projects.

You can also use it as non Rx, my application does it and it's doable, since the doc also improved you should definitely give it a try before complaining (in my opinion), if you didn't already.

The complaint here is for MS not doing what the competition is doing, I think. From my experience in simple BT scenarios (which is 90% of the scenarios you usually have in mobile apps) the 3rd party libs we have are pretty decent.

**Reply by alunmj** [reply]
The Shiny documentation, particularly for Bluetooth LE, is awful. Hard to read, incomplete, inconsistent, occasionally contradictory.
Even just reading the samples and coding to that is difficult.
There appears to be no BLE support in Shiny for UWP, because it crashes - hard - just trying to call Resolve() on the first Bluetooth interface.
Needles to say, this isn't mentioned at all in the documentation.
There should be a clear, simple to use, well-documented and cross-platform (so, all platforms supported by MAUI that have a BLE stack) BLE implementation as a part of MAUI.

**Reply by nissaba** [reply]
Funny Rad Studio has had a Bluetooth/BLE plugin for ages now.

**Reply by ChumKiu** [reply]
> BLE is not on our roadmap for .NET 6. We are discussing bringing Essentials together with Forms in dotnet/maui as there's a growing overlap and interdependence.
> 
> Still, BLE is such a complex topic I'm not sure it'll ever be something we bake into this product. [Shiny](https://github.com/shinyorg/shiny) seems to be filling this need presently.

It is seems to ma a nonsense because it will get difficult develop app for IoT based on BLE technology, I personally as hw designer and software programmer for my devices hate such thinking because it will not follow what is the common path of the technology. If .NET MAUI it is thinked to glue in one place all the code to deploy on many devices it is not be aware of peripheral usage on such devices and Bluetooth is almost present in every device now, I think is limiting.

Just my 2 cents

BR
ChumKiu

**Reply by aritchie** [reply]
The level of entitlement in this thread is insane.  I came in to offer some help, but wow...no thanks... the behaviour here is unprofessional and just plain rude. Microsoft doesn't have to offer this feature.  Build it yourself... no other cross platform solution offers stuff like this out of the box.

@alunmj + all the thumbs up to your thread - I owe you nothing.  I build Shiny for free.  To go to this level of public complaint about docs, samples, etc shows a real poor attitude.  I don't recall seeing anything you've put out for the community... nor do I recall a request for help.... an offer to help.... or a sponsorship.  I challenge you to build better and put it out there for others.  

@LITTOMA RX is quite appropriate for BLE.  The biggest native implementations for Android & iOS both use RX.

**Reply by alunmj** [reply]
There's no entitlement intended here.
Nor am I intending to be professional on a purely personal, hobby, project.
I'm sorry you think that I came here to join a community or contribute code. I'm just trying to put lights on a toy.
I came here looking for a way to move an app from Xamarin to MAUI. So, although I'm a developer, in relation to your framework I'm a user.
Without any malice at all, I'm not looking to help write documentation for a novel model that I don't remotely understand - the library I use now doesn't make me call "Resolve" before I can call the first function I need - and you give the impression that you'd be just as dismissive of any documentation assistance I might give.
It's truly impressive that you've put together an entire framework and support it for the community. That's great. I don't want the framework. Or the community.
I only want BLE support.
That's it.
When I asked how to get BLE support on MAUI, the overwhelming answer I got was "Why don't you just drop Shiny in your project?"
And when I tried to, the answer was clearly "because it's not just a drop in, you have to puzzle your way through understanding so much source code and extraneous platforming that you're expected to become part of the community just to get the basics to work".
I'll stick with the library I've been using, and wait until there's a simple way to include BLE in a MAUI project before upgrading from Xamarin.Forms.
As to whether Microsoft "owes" me BLE support? I guess they owe me nothing, although goodness knows I've spent enough with them over the last thirty years of developing against, and securing, their various platforms and SDKs. As far as communities go, I've been an MVP for at least one of those decades.
No, they don't owe me anything, but if they want me to use MAUI, instead of slouching reluctantly into learning the Android platform, adding enough code to MAUI to support BLE seems like a relatively small piece of work.
Nobody has to do anything for me. But what they do will change where I go.
I'm not entitled. I'm goal-oriented.

**Reply by marciosouzajunior** [reply]
Just want to add that I also looked for a solution and between the options, [this one](https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le) worked pretty well in my project. And also that, is such a pain to accomplish anything in Xamarin. Because of that I'm moving to React in new projects.

**Reply by ChrisKarle** [reply]
It's now entitlement to ask for a common API to a ubiquitous network protocol to be included in a product Microsoft _wants_ me to use?  C'mon, grow up.  Yeah, let me start coding up my own IP stack so I don't feel entitled asking for that.

**Reply by aritchie** [reply]
It's how you ask it.  If you keep complaining that Microsoft won't do something for you after they've already said "we're not doing it", that truly is entitlement.  MAUI is already free and open source.  The team is known to be understaffed.  They can spend their time better elsewhere.  They also get to decide where their time is spent best since they're the ones funding the project.  I need something for my business that isn't available in a form I need, I write it and in a lot of cases, I give it away as open source for others to potentially benefit.  But you do you - keep complaining, offering nothing to improve the situation, and telling people to "grow up"

**Reply by nexxuno** [reply]
And yet again keep asking for something to a big corporation, i e. Showing interest, is the only way to hope for some first party plugin. We are asking  Microsoft, I don't think they are so touchy about how you ask things, if they are maybe they shouldn't be the ones deciding priorities. They should just gather interest and decide priorities, who cares if someone is not nice enough...

Your libraries are good but some official support would be better (and could be built on what you already developed).

**Reply by ChrisKarle** [reply]
I have reread this thread several times.  I am absolutely struggling to see what you found inappropriate.  Several people articulated their position that BLE support should be included in MAUI, and one person called the documentation for your BLE library as "awful".  Instead of asking the commenter what they feel you could improve or saying "sorry, I do this for free, but you can pay me to help you because I need to keep my lights on too", you decided to call everyone asking for BLE support as entitled, unprofessional, and rude (which nobody was being any of those things)... and you continued to further berate the particular commenter criticizing your documentation.  You also throw in a little drama/hyperbole too, "to go to this level of public complaint... shows a real poor attitude".  Your post served no other purpose than to claim you could help people, called everyone names, then you took your ball and went home.  I do not find that to be mature behavior.

Let me clarify.  I am telling *you* to "grow up".  I apologize if it sounds harsh, but I cannot think of something better.  I do not care what reasons that Microsoft has for not adding BLE support to MAUI.  Again, let us be adults here.  Microsoft is a business; they do not do *anything* out of the kindness of their heart.  If they want me to use MAUI (thus increasing their influence), they need to offer something I can use to get my work done better or cheaper than someone else.  This included not creating dependencies on external projects that can stopped being maintained (always fun to update a project in 5 years).  I also do not care if someones project becomes irrelevant because Microsoft natively adds its functionality to MAUI.  This is the adult world, and I need to get stuff done.  There is no need to berate people when you receive some criticism regarding your project.  There is also no reason to call people entitled, unprofessional, and rude for simply asking for something they need to get their work done.  If we keep asking, maybe it's important and Microsoft should listen.

**Reply by aritchie** [reply]
> you called everyone names

Nope.  Just a select few whether they know they are acting entitled or not, is I guess a point of view.  Be nice, be kind, be constructive - this is a free product with a very limited team who are under a lot of stress already.

> I am absolutely struggling to see what you found inappropriate

It is easy for you to take no offence.  You haven't put anything out publicly that I can see.  I didn't ask to have my stuff recommended, nor did I ask for feedback.  I give it away in hopes that it helps people and I put quite a bit of work into it, but there are comments like "awful", "wrong patterns", etc. all from people who seemingly consume and not give back.  What an age we've come to where people complain about free things.  Don't use it then.... 

I see people saying how MAUI isn't really cross platform because Microsoft won't built BLE (right now anyhow).  Again, it's how you ask.  +1's aren't the issue.   I don't think you will understand this standpoint, so we may as well agree to disagree at this point.

> This is the adult world, and I need to get stuff done

There is no official support from any other cross platform framework (Flutter, React Native, etc).  You will be in the same boat no matter what xplat platform you go to.  +1 the topic and move on.

> Let me clarify.  I am telling you to "grow up". I apologize if this sounds harsh

I don't need your apology or your recommendation.  I don't know you so it doesn't carry any weight.  I would hope people in general see why I'm saying what I'm saying.  Some definitely do and some don't.  That's fine.  

At this point, let's be adults and agree to disagree.  You have no public profile that I can see here on github, so you have the satisfaction of your words not really carrying weight outside of this very anonymous set of posts.

**Reply by ChrisKarle** [reply]
@aritchie You seem to labor under the assumption that you have been somehow granted certain authority or privilege.  

1.) People can make comments regarding your project without requiring permission or solicitation from you.  The commenter called your documentation "awful".  Yes, it would have been better if they cited examples or even created a pull request, but this does not excuse your behavior.

2.) I do not understand your focus on the number of public repositories that I or someone else may have.  It has no relevance regarding my petition to add a native Bluetooth API to MAUI, nor does it prohibit me from criticizing your behavior.  Your opinion on such matters does not sit higher than anyone else's.

3.) Again, I do not care the reasons that some functionality is not present.  I do not mean this to be callous, as we all have our priorities.  We are not pestering some 1 or 2 person development team.  We are asking a 2 trillion dollar company with 220,000 employees.

If it was not for your behavior, I would have simply +1 the topic and moved on.  Now, come back with whatever final word that you feel is important.  I'm done here.

**Comment by salarcode** [comment]
Meanwhile we can use libraries, here is a one that recently got MAUI support:

[InTheHand.BluetoothLE](https://www.nuget.org/packages/InTheHand.BluetoothLE/) and [InTheHand.Net.Bluetooth](https://www.nuget.org/packages/InTheHand.Net.Bluetooth)
Source: https://github.com/inthehand/32feet

**Reply by LITTOMA** [reply]
good enough.

**Comment by gfmoore** [comment]
Not really because until you can set bluetooth permissions it's useless. And guess what, there is currently no code to do that in Maui.

**Reply by salarcode** [reply]
We need to define custom permissions for each platform separately. Have a look here:
https://github.com/dotnet/maui/tree/main/src/Essentials/src/Permissions

Update: check the git gist in comment below

**Reply by LITTOMA** [reply]
The 32feet library cannot discover any BLE device on a Andorid device. Not even I have granted permissions.

**Reply by jfversluis** [reply]
Just because something is not (yet) surfaced in .NET MAUI doesn't mean you can't do it. 

1. You can implement your own way of requesting the Bluetooth permissions
2. The current Permissions API is pretty extensible and you can just add your own to that: https://learn.microsoft.com/dotnet/maui/platform-integration/appmodel/permissions?view=net-maui-7.0&tabs=ios#extending-permissions

**Comment by salarcode** [comment]
I have created a custom permission for android which follows the specs about permissions.
Hope this helps
https://gist.github.com/salarcode/da8ad2b993e67c602db88a62259d0456

**Reply by LITTOMA** [reply]
nice

**Reply by gfmoore** [reply]
Not only nice, but  brilliant. I did have a couple of queries which I've posted on your code page, but honestly friend you will have made many people very happy now and as Maui get used more in the future. Why MS couldn't have implemented this....

But how did you know how to do this? I followed the documentation https://docs.microsoft.com/en-us/dotnet/maui/platform-integration/appmodel/permissions?tabs=android#extending-permissions
and could get the location part working, but could see that somehow code needed to be written using that template, but I had no idea how to do this.

**Reply by salarcode** [reply]
@gfmoore I had a look here where I could see how other permissions are implemented from MAUI team: https://github.com/dotnet/maui/tree/main/src/Essentials/src/Permissions
I have replied to you on the github gist page about other question.

**Reply by gfmoore** [reply]
Cheers :)

**Comment by TimoKinnunen** [comment]
Bluetooth LE is a headache, take a look here https://github.com/TimoKinnunen/MauiXAMLBluetoothLE. This is a good starting point to bring Bluetooth LE (Plugin.BLE) closer to .Net MAUI, I think.

**Comment by orwo1** [comment]
@davidortinau @jfversluis 
Is there any update on this issue?
Is there going to be a Bluetooth permission built in Maui?

**Reply by jfversluis** [reply]
Probably this comment from this same discussion is relevant here: https://github.com/dotnet/maui/discussions/277#discussioncomment-4631312

Additionally in the meantime, this one was merged so they should be available in .NET 8: https://github.com/dotnet/maui/pull/12413

**Reply by orwo1** [reply]
The content merged in .Net 8, is it only Android Bluetooth permission,
or will it be cross platform and support iOS as well?

**Reply by jfversluis** [reply]
Just the permission. We don't have any plans to include this on the roadmap for .NET 8

**Comment by yshteinm** [comment]
Just a note. We kept telling Microsoft we need WCF in .NET Core. And here we are-in few years we have support. May be that will work with Bluetooth too.

**Comment by filipoff2** [comment]
Guys manage your expectations :) 

I would be extremely happy if team could focus on multi platform app UI, a specially on UI part which is name 🗡️  :P 
Not more and not less. 

Thanks

**Comment by gfmoore** [comment]
My expectations are that I will not be using maui. It's not fit for my purpose. Sad really. :(

**Reply by TimoKinnunen** [reply]
Android operating system does not have a uniform way of granting permissions through it's history of versions. And that has technicians at Microsoft to handle in some way. And Microsoft does amazing work with MAUI!
